src/components/accordion/Accordion.jsx

```diff
@@ -80,11 +94,86 @@ const Accordion = ({
   className = '',
   spacing = 's',
   reduceMotion = false,
+  defaultExpanded,
+  expanded,
+  onChange,
 }: AccordionPropsType) => {
   const wrapperRef = useRef<HTMLDivElement | null>(null);
-  const [state, dispatch] = useReducer(reducer, {
-    opened: {},
-    focusedElementId: null,
+  const isControlled = expanded !== undefined;
+  const {current: wasControlled} = useRef<boolean>(isControlled);
+  const isCallbackMissing = isControlled && !onChange;
+  const hasComponentChangedToUncontrolled = wasControlled && !isControlled;
+  const hasComponentChangedToControlled = !wasControlled && isControlled;
+  const isAllowMultiplePassedForControlled = isControlled && allowMultiple;
+
+  if (__DEV__) {
+    invariant(
+      !isCallbackMissing,
+      // eslint-disable-next-line max-len
+      ' You provided an `expanded` prop to a Accordion without an `onChange` handler. Users won`t be able to switch between expanded/collapsed state.'
+    );
+
+    invariant(
+      !hasComponentChangedToUncontrolled,
+      'You cannot change Accordion component from controlled to uncontrolled variant.'
+    );
+
+    invariant(
+      !hasComponentChangedToControlled,
+      'You cannot change Accordion component from uncontrolled to controlled variant.'
+    );
+
+    invariant(
+      !isAllowMultiplePassedForControlled,
+      'allowMultiple is not working in controlled Accordion'
+    );
+
+    invariant(
+      !(
+        !allowMultiple &&
+        Array.isArray(defaultExpanded) &&
+        defaultExpanded.length > 1
+      ),
+      // eslint-disable-next-line max-len
+      'defaultExpanded is an array with more than 1 element but allowMultiple prop is not set. The first value from the array was picked as a default expanded. Set allowMultiple attribute or provide only one default expanded.'
+    );
+  }
+
+  const getUpdatedOpenedItems = useCallback(
+    (expanded: ExpandedItemsType, id: string, value: boolean) => {
+      return allowMultiple ? {...expanded, [id]: value} : {[id]: value};
+    },
+    [allowMultiple]
+  );
+
+  const [state, dispatch] = useReducer(reducer, null, () => {
+    if (isControlled) {
+      return {
+        expanded: {},
+        focusedElementId: null,
+      };
+    }
+
+    if (defaultExpanded !== undefined) {
+      const expandedArray = !Array.isArray(defaultExpanded)
+        ? [defaultExpanded]
+        : defaultExpanded;
```

Comment:
I think we can skip negation in this condition
```suggestion
      const expandedArray = Array.isArray(defaultExpanded)
        ? defaultExpanded
        : [defaultExpanded];
```